README.md
# <h1 align="center">reth-db-py</h1>

**Python package allowing you to interact with the Reth DB via Python. Written with Rust and Pyo3.**

**This python wrapper can access node data 3x-5x faster than local web3.py calls.  Using this package, a
block can be retrieved in ~10ms on a local reth DB.**

[![Test Suite](https://github.com/gibz104/reth-db-py/actions/workflows/CI.yml/badge.svg)](https://github.com/gibz104/reth-db-py/actions/workflows/CI.yml)
[![Py Versions](https://img.shields.io/badge/python-3.7_|_3.8_|_3.9_|_3.10-blue.svg)](https://www.python.org/downloads/)
[![Test OS](https://img.shields.io/badge/tested_on-ubuntu_|_mac_os-blue.svg)](https://github.com/gibz104/google-sheets-writer/actions/workflows/tests.yaml)

# Installation
This package has been published to PyPi and can be installed using pip:
```bash
pip install reth-db-py
````



# Assets
This package only has a single python class made available: `PyDatabaseHandler`.  

`PyDatabaseHandler` is a class used to interact with the Reth DB.  It's a wrapper around the Rust `DatabaseHandler` struct.
It has a few methods which all return json strings:
* `get_header_by_block_number`: get a single header by block number
* `get_headers_by_block_number_range`: get multiple headers by block number range
* `get_transaction_by_id`: get a single transaction by transaction id
* `get_transactions_by_id_range`: get multiple transactions by transaction id range
* `get_transactions_by_block_number_range`: get multiple transactions by block number range
* `get_block_by_number`: get a single block by block number
* `get_uncles_by_block_number`: get uncles by block number
* `get_receipts_by_transaction_id`: get receipts by transaction id
* `get_receipts_by_block_number`: get receipts by block number

```rust
impl PyDatabaseHandler {
    pub fn get_header_by_block_number(&self, number: u64) -> PyResult<String>
    pub fn get_headers_by_block_number_range(&self, start: u64, end: u64) -> PyResult<String>
    pub fn get_transaction_by_id(&self, id: u64) -> PyResult<String>
    pub fn get_transactions_by_id_range(&self, start: u64, end: u64) -> PyResult<String>
    pub fn get_transactions_by_block_number_range(&self, start: u64, end: u64) -> PyResult<String>
    pub fn get_block_by_number(&self, number: u64) -> PyResult<String>
    pub fn get_uncles_by_block_number(&self, number: u64) -> PyResult<String>
    pub fn get_receipts_by_transaction_id(&self, id: u64) -> PyResult<String>
    pub fn get_receipts_by_block_number(&self, number: u64) -> PyResult<String>
}
```

```python
class PyDatabaseHandler:
    def get_header_by_block_number(self, number: int) -> str
    def get_headers_by_block_number_range(self, start: int, end: int) -> str
    def get_transaction_by_id(self, id: int) -> str
    def get_transactions_by_id_range(self, start: int, end: int) -> str
    def get_transactions_by_block_number_range(self, start: int, end: int) -> str
    def get_block_by_number(self, number: int) -> str
    def get_uncles_by_block_number(self, number: int) -> str
    def get_receipts_by_transaction_id(self, id: int) -> str
    def get_receipts_by_block_number(self, number: int) -> str
```




# Usage
#### Import reth-db-py assets:
```python
from reth_db_py import PyDatabaseHandler
```

#### Create a PyDatabaseHandler instance:
```python
handler = PyDatabaseHandler("/path/to/db/mdbx.dat")
```

#### Get the header by block number
```python
header = handler.get_header_by_block_number(17_000_000)
```

#### Get the headers by block number range
```python
headers = handler.get_headers_by_block_number_range(17_000_000, 17_000_005)
```

#### Get transaction by id
```python
transaction = handler.get_transaction_by_id(1000)
```

#### Get transactions by id range
```python
transactions = handler.get_transactions_by_id_range(1000, 1005)
```

#### Get transactions by block number range
```python
transactions = handler.get_transactions_by_block_number_range(17_000_000, 17_000_005)
```

#### Get block by number
```python
block = handler.get_block_by_number(17_000_000)
```

#### Get uncles by block number
```python
uncles = handler.get_uncles_by_block_number(17_000_000)
```

#### Get receipts by transaction id
```python
receipts = handler.get_receipts_by_transaction_id(1000)
```

#### Get receipts by block number
```python
receipts = handler.get_receipts_by_block_number(17_000_000)
```

# Tests
Coming soon.


# Benchmarks
Speed tests were conducted by retrieving 1,000 block bodies (including transactions)
using: 
* reth-db-py (local reth db)
* web3.py (local reth JSON RPC)
* http requests (remote alchemy API)

See benchmark tests [here](examples/benchmarks.py).

![benchmarks](examples/assets/benchmarks.png)


CHANGELOG.md
[//]: # ("Added", "Changed", "Deprecated", "Removed", "Fixed", and "Security")

## [0.1.4] - 2023-07-05

### Added
- Python type annotations in reth_db_py.pyi
- Added class `PyDatabaseHandler`
- Added methods `get_header_by_block_number`, `get_headers_by_block_number_range`, `get_transaction_by_id`,
  `get_transactions_by_id_range`, `get_transactions_by_block_number_range`, `get_block_by_number`,
  `get_uncles_by_block_number`, `get_receipts_by_transaction_id`, and `get_receipts_by_block_number`.
- Added changelog

### Removed
- Removed `DbHandler` class
- Removed `TableName` enum

### Changed
- Switched to using reth providers to access database

## [0.1.3] - 2023-06-29

### Added
- Initial release of the project.



src/types.rs
use crate::utils::DatabaseHandler;
use pyo3::prelude::*;
use pyo3::exceptions;


#[pyclass]
pub struct PyDatabaseHandler {
    provider: DatabaseHandler,
}

#[pymethods]
impl PyDatabaseHandler {
    #[new]
    pub fn new(db_path: &str) -> PyResult<Self> {
        match DatabaseHandler::new(db_path.to_string()) {
            Ok(provider) => Ok(Self { provider }),
            Err(err) => Err(PyErr::new::<exceptions::PyException, _>(format!("{:?}", err))),
        }
    }

    pub fn get_header_by_block_number(&self, number: u64) -> PyResult<String> {
        match self.provider.get_header_by_block_number(number) {
            Ok(header) => Ok(header),
            Err(err) => Err(PyErr::new::<exceptions::PyException, _>(format!("{:?}", err))),
        }
    }

    pub fn get_headers_by_block_number_range(&self, start: u64, end: u64) -> PyResult<String> {
        match self.provider.get_headers_by_block_number_range(start, end) {
            Ok(headers) => Ok(headers),
            Err(err) => Err(PyErr::new::<exceptions::PyException, _>(format!("{:?}", err))),
        }
    }

    pub(crate) fn get_transaction_by_id(&self, id: u64) -> PyResult<String> {
        match self.provider.get_transaction_by_id(id) {
            Ok(transaction) => Ok(transaction),
            Err(err) => Err(PyErr::new::<exceptions::PyException, _>(format!("{:?}", err))),
        }
    }

    pub(crate) fn get_transactions_by_id_range(&self, start: u64, end: u64) -> PyResult<String> {
        match self.provider.get_transactions_by_id_range(start, end) {
            Ok(transactions) => Ok(transactions),
            Err(err) => Err(PyErr::new::<exceptions::PyException, _>(format!("{:?}", err))),
        }
    }

    pub(crate) fn get_transactions_by_block_number_range(&self, start: u64, end: u64) -> PyResult<String> {
        match self.provider.get_transactions_by_block_number_range(start, end) {
            Ok(transactions) => Ok(transactions),
            Err(err) => Err(PyErr::new::<exceptions::PyException, _>(format!("{:?}", err))),
        }
    }

    pub(crate) fn get_block_by_number(&self, number: u64) -> PyResult<String> {
        match self.provider.get_block_by_number(number) {
            Ok(block) => Ok(block),
            Err(err) => Err(PyErr::new::<exceptions::PyException, _>(format!("{:?}", err))),
        }
    }

    pub(crate) fn get_uncles_by_block_number(&self, number: u64) -> PyResult<String> {
        match self.provider.get_uncles_by_block_number(number) {
            Ok(uncles) => Ok(uncles),
            Err(err) => Err(PyErr::new::<exceptions::PyException, _>(format!("{:?}", err))),
        }
    }

    pub(crate) fn get_receipts_by_transaction_id(&self, id: u64) -> PyResult<String> {
        match self.provider.get_receipts_by_transaction_id(id) {
            Ok(receipts) => Ok(receipts),
            Err(err) => Err(PyErr::new::<exceptions::PyException, _>(format!("{:?}", err))),
        }
    }

    pub(crate) fn get_receipts_by_block_number(&self, number: u64) -> PyResult<String> {
        match self.provider.get_receipts_by_block_number(number) {
            Ok(receipts) => Ok(receipts),
            Err(err) => Err(PyErr::new::<exceptions::PyException, _>(format!("{:?}", err))),
        }
    }
}


src/utils.rs
use reth_db::open_db_read_only;
use reth_primitives::ChainSpecBuilder;
use reth_provider::{BlockReader, HeaderProvider, ProviderFactory, ReceiptProvider, TransactionsProvider};

use std::path::Path;
use std::sync::Arc;
use reth_db::DatabaseEnvRO;

pub struct DatabaseHandler {
    factory: ProviderFactory<DatabaseEnvRO>,
}

impl DatabaseHandler {
    pub fn new(db_path: String) -> eyre::Result<Self> {
        let db = open_db_read_only(&Path::new(&db_path), None)?;
        let spec = Arc::new(ChainSpecBuilder::mainnet().build());
        let factory = ProviderFactory::new(db, spec.clone());
        Ok(Self { factory })
    }

    pub(crate) fn get_header_by_block_number(&self, number: u64) -> eyre::Result<String> {
        let provider = self.factory.provider()?;
        let header = provider.header_by_number(number)?;
        let json = serde_json::to_string(&header)?;
        Ok(json)
    }

    pub(crate) fn get_headers_by_block_number_range(&self, start: u64, end: u64) -> eyre::Result<String> {
        let provider = self.factory.provider()?;
        let headers = provider.sealed_headers_range(&start..&end)?;
        let json = serde_json::to_string(&headers)?;
        Ok(json)
    }

    pub(crate) fn get_transaction_by_id(&self, id: u64) -> eyre::Result<String> {
        let provider = self.factory.provider()?;
        let transaction = provider.transaction_by_id(id)?;
        let json = serde_json::to_string(&transaction)?;
        Ok(json)
    }

    pub(crate) fn get_transactions_by_id_range(&self, start: u64, end: u64) -> eyre::Result<String> {
        let provider = self.factory.provider()?;
        let transactions = provider.transactions_by_tx_range(&start..&end)?;
        let json = serde_json::to_string(&transactions)?;
        Ok(json)
    }

    pub(crate) fn get_transactions_by_block_number_range(&self, start: u64, end: u64) -> eyre::Result<String> {
        let provider = self.factory.provider()?;
        let transactions = provider.transactions_by_block_range(&start..&end)?;
        let json = serde_json::to_string(&transactions)?;
        Ok(json)
    }

    pub(crate) fn get_block_by_number(&self, number: u64) -> eyre::Result<String> {
        let provider = self.factory.provider()?;
        let block = provider.block(number.into())?;
        let json = serde_json::to_string(&block)?;
        Ok(json)
    }

    pub(crate) fn get_uncles_by_block_number(&self, number: u64) -> eyre::Result<String> {
        let provider = self.factory.provider()?;
        let uncles = provider.ommers(number.into())?;
        let json = serde_json::to_string(&uncles)?;
        Ok(json)
    }

    pub(crate) fn get_receipts_by_transaction_id(&self, id: u64) -> eyre::Result<String> {
        let provider = self.factory.provider()?;
        let receipts = provider.receipt(id)?;
        let json = serde_json::to_string(&receipts)?;
        Ok(json)
    }

    pub(crate) fn get_receipts_by_block_number(&self, number: u64) -> eyre::Result<String> {
        let provider = self.factory.provider()?;
        let receipts = provider.receipts_by_block(number.into())?;
        let json = serde_json::to_string(&receipts)?;
        Ok(json)
    }
}


examples/benchmarks.py
import timeit
import seaborn as sns
import pandas as pd
import matplotlib.pyplot as plt
import statistics
from web3 import Web3
import requests
import json
from reth_db_py import PyDatabaseHandler


def get_block_local_web3(block_number: int):
    # Connect to Ethereum node using Infura
    w3 = Web3(Web3.HTTPProvider("http://127.0.0.1:8545"))
    # Get block data
    _block = w3.eth.get_block(block_number, full_transactions=True)


def get_block_alchemy(block_number: int):
    url = "https://eth-mainnet.g.alchemy.com/v2/zpAe-NVcByLAtwahyJUvoSdpAHcmDEqr"
    headers = {"Content-Type": "application/json"}
    # The method eth_getBlockByNumber gets the block data by its number
    data = {
        "jsonrpc": "2.0",
        "id": 1,
        "method": "eth_getBlockByNumber",
        "params": [hex(block_number), True]
    }
    # Get block data
    _response = requests.post(url, headers=headers, data=json.dumps(data)).json()


def get_block_reth_db_py(block_number: int):
    # Path to the MDBX database file (created by Reth process)
    handler = PyDatabaseHandler("/mnt/Node_Data/docker/volumes/local_reth_nodedata/_data/reth/db/mdbx.dat")
    # Get block data
    _block = handler.get_block_by_number(block_number)


functions = [get_block_local_web3, get_block_alchemy, get_block_reth_db_py]
num_runs = 1_000
results = []

start_block = 16_000_000
for function in functions:
    function_name = function.__name__
    run_times = [timeit.timeit(lambda: function(start_block + i), number=1) for i in range(num_runs)]
    average_run_time = sum(run_times) / len(run_times)
    std_dev_run_time = statistics.stdev(run_times)

    print(f"Function: {function_name}")
    print(f"Average run time over {num_runs} runs: {average_run_time * 1000} milliseconds")
    print(f"Standard deviation of run time: {std_dev_run_time * 1000} milliseconds")
    print("\n")

    # Modify function name to include average run time
    function_name += f' (Avg: {average_run_time * 1000:.2f} ms)'

    # Add to results
    for i, run_time in enumerate(run_times):
        results.append(
            {'Function': function_name, 'Run Time': run_time * 1000, 'Run Count': i+1})

# Create DataFrame from results
df = pd.DataFrame(results)

# Plotting
sns.lineplot(data=df, x='Run Count', y='Run Time', hue='Function')
plt.yscale('log')  # Set y-axis to logarithmic scale
plt.ylabel('Run Time (ms)')  # Update y-axis label
plt.title('Run Times Distribution')
plt.show()


examples.py
# Import the DbHandler and TableName from reth_db_py
from reth_db_py import PyDatabaseHandler
from time import perf_counter_ns
import requests
import json


# Path to the MDBX database file (created by Reth process)
handler = PyDatabaseHandler("/mnt/Node_Data/docker/volumes/local_reth_nodedata/_data/reth/db/mdbx.dat")


# Get the header by block number
header = handler.get_header_by_block_number(17_000_000)

# Get the headers by block number range
headers = handler.get_headers_by_block_number_range(17_000_000, 17_000_005)

# Get transaction by id
transaction = handler.get_transaction_by_id(1000)

# Get transactions by id range
transactions = handler.get_transactions_by_id_range(1000, 1005)

# Get transactions by block number range
transactions = handler.get_transactions_by_block_number_range(17_000_000, 17_000_005)

# Get block by number
block = handler.get_block_by_number(17_000_000)

# Get uncles by block number
uncles = handler.get_uncles_by_block_number(17_000_000)

# Get receipts by transaction id
receipts = handler.get_receipts_by_transaction_id(1000)

# Get receipts by block number
receipts = handler.get_receipts_by_block_number(17_000_000)



# HTTP RPC Call vs. Reth-db-py call
url = "http://127.0.0.1:8545"
headers = {
    'Content-Type': 'application/json',
}
payload = {
    "method": "eth_blockNumber",
    "params": [],
    "id": 1,
    "jsonrpc": "2.0"
}
start_time = perf_counter_ns()
response = requests.post(url, headers=headers, data=json.dumps(payload))
duration_rpc = (perf_counter_ns() - start_time) / 1_000_000.0
print(f'RPC time: {duration_rpc} ms')

# Reth-db-py call to get block hash
start_time = perf_counter_ns()
header_hash = handler.get_header_by_block_number(17_000_000)
duration = (perf_counter_ns() - start_time) / 1_000_000.0
print(f'reth_db_py time: {duration} ms')

print(f'reth-db-py is {duration_rpc / duration} times faster than RPC')
